akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2ServerDemux.scala

```diff
@@ -9,25 +9,44 @@ import akka.http.impl.engine.http2.FrameEvent._
 import akka.http.impl.engine.http2.Http2Protocol.ErrorCode
 import akka.http.impl.engine.http2.Http2Protocol.ErrorCode.FLOW_CONTROL_ERROR
 import akka.http.impl.engine.http2.Http2Protocol.SettingIdentifier
+import akka.http.scaladsl.model.AttributeKey
+import akka.http.scaladsl.model.HttpEntity.{ ChunkStreamPart, LastChunk }
+import akka.http.scaladsl.model.headers.RawHeader
 import akka.http.scaladsl.settings.Http2CommonSettings
 import akka.macros.LogHelper
 import akka.stream.Attributes
 import akka.stream.BidiShape
 import akka.stream.Inlet
 import akka.stream.Outlet
 import akka.stream.impl.io.ByteStringParser.ParsingException
+import akka.stream.scaladsl.Source
 import akka.stream.stage.GraphStage
 import akka.stream.stage.GraphStageLogic
 import akka.stream.stage.InHandler
 import akka.stream.stage.StageLogging
 import akka.util.ByteString
 
 import scala.collection.immutable
+import scala.concurrent.{ ExecutionContext, Future }
 import scala.util.control.NonFatal
 
 /** Currently only used as log source */
 @InternalApi
-private[http2] sealed abstract class Http2ClientDemux
+private[http2] class Http2ClientDemux(http2Settings: Http2CommonSettings, initialRemoteSettings: immutable.Seq[Setting])
+  extends Http2Demux[ChunkedHttp2SubStream](http2Settings, initialRemoteSettings, upgraded = false, isServer = false) {
+
+  override def createSubstream(initialHeaders: ParsedHeadersFrame, data: Source[ByteString, Any], trailingHeaders: Future[ParsedHeadersFrame], correlationAttributes: Map[AttributeKey[_], _])(implicit ec: ExecutionContext): ChunkedHttp2SubStream =
+    ChunkedHttp2SubStream(initialHeaders, data.map(ChunkStreamPart(_)).concat(Source.lazilyAsync(() => trailingHeaders.map(frame => LastChunk(extension = "", frame.keyValuePairs.map {
```

Comment:
That feels like a hack. I guess since it's a somewhat non-essential part, it might be good enough? Maybe add a comment about the reasoning behind that implementation.

Alternatively, could the logic to create the Http2SubStream be part of `IncomingStreamState` and emit the right elements there right away? I guess that would be more invasive because of the different element types that `IncomingStreamBuffer` would have to support.

Reply:
Yes, that's the dilemma I meant in the PR message ;). I'll first fix up this PR and then have another go at coming up with an alternative approach.